### PR TITLE
Add a transport socket match in cluster.

### PR DESCRIPTION
**Pull request by incfly**
Signed-off-by: Jianfei Hu <jianfeih@google.com>

API for https://github.com/envoyproxy/envoy/issues/8016

Context

Customers adopting service mesh likes mTLS ability. However, rolling it out without breaking existing traffic is hard. This is because mTLS is configured on per cluster basis. In reality, a service consists of multiple endpoints, mixed with having Envoy sidecar and without-sidecar endpoints. Client envoy can't send mTLS traffic until all server migrated to having Envoy sidecar.

This API tries to solve the issue by allowing mTLS/transport socket to configured at finer granularity, e.g. endpoint level. The endpoint has metadata label information, which will be used to decide which transport socket configuration to use from a map specified in the cluster.

So the outcome is that, xDS management server is able to configure client envoy talks to endpoints with sidecar in mTLS and plain text to endpoints without sidecar, for a single cluster.

Description: 
Risk Level: N/A (API change only)
Release Notes: Cluster API change to use different transport socket based on endpoint label.

### CODE REVIEW

File: api/envoy/api/v2/cds.proto [hunk 1]
```diff
@@ -234,6 +234,32 @@ message Cluster {
   //   verification.
   auth.UpstreamTlsContext tls_context = 11;
 
+  // UpstreamTransportSocket specifies the transport socket config used for upstream connection.
+  message UpstreamTransportSocket {
```
Comment (incfly) [inline api/envoy/api/v2/cds.proto]: Maybe this should be used for default `transport_socket` `tls_context` in the cluster as well? Looks clearer to me to use the same proto msg, but that requires xDS management server changes to migrate.

Not sure if it's worth it.

File: api/envoy/api/v2/cds.proto [hunk 2]
```diff
@@ -234,6 +234,32 @@ message Cluster {
...
+  //
+  //   {
+  //     "envoy.transport_socket_matcher_label":
+  //        { "label": "mtlsEnabled" }
+  //   }
+  //
+  // This endpoint will use the transport_socket_matcher["mtlsEnabled"] configuration.
+  // In case of of label is missing or the transport_socket_matcher does not have corresponding
+  // entry in the map, Envoy fallbacks to the *tls_context* or *transport_socket* in the cluster.
+  map<string, UpstreamTransportSocket> transport_socket_matcher = 41;
```
Comment (incfly) [inline api/envoy/api/v2/cds.proto]: I'm hesitating of terminology, `selector` might be a better term?

File: api/envoy/api/v2/cds.proto (hunk 1 above)
Comment (lizan) [reply api/envoy/api/v2/cds.proto]: That's why I filed #8098, then you shouldn't need the oneof below.

File: api/envoy/api/v2/cds.proto (hunk 2 above)
Comment (lizan) [reply api/envoy/api/v2/cds.proto]: This seems inconsistent from the proposal doc in #8016, why this is a map? cc @PiotrSikora

Comment (incfly) [reply api/envoy/api/v2/cds.proto]: okay, just realized the doc is using the repeated list.

In that case, we should define `UpstreamTransportSocket` to include a field like `string match`.

And then replace this map with

`repeated UpstreamTransportSocket transport_socket_matcher = 41`.

Does that sound better? I can't tell how it's better than map or any issue with the map though. I mean the internal implementation we still need to have a map for lookup.

Comment (incfly) [reply api/envoy/api/v2/cds.proto]: Take another look

We already have the [`TransportSocket`](https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/core/base.proto#L218)
defined.

We already have implementation to the `tls_context` to this struct in [`upstream_impl`](https://sourcegraph.com/github.com/envoyproxy/envoy@master/-/blob/source/common/upstream/upstream_impl.cc?suggestion#L727)

Maybe we can just use define it as

```proto
message Cluster {
  message UpstreamTransportSocketMatch {
     string endpoint_label_match;
     core.TransportSocket transport_socket;
  }
  repeated UpstreamTransportSocketMatch transport_socket_matcher;
}
```

Then the xDS server thrust `TlsContext` config into the `core.TransportSocket.Config` typed as struct.Any.

Basically I guess same as what proposed in #8098 get rid of the oneof.

Comment (htuch) [inline api/envoy/api/v2/cds.proto]: Can we just put this in EDS itself? I think I'm missing the design discussion here. Also, how does this relate to the aggregate cluster feature?

**Comment by htuch** [comment]
I'm generally opposed to using unstructured metadata for this purpose. If we are going to go down this path (see above comment about needing to understand design rationale and motivation), then we need to add an explicit label field to EDS endpoints.

**Comment by htuch** [comment]
@incfly FYI, for changes to Envoy's APIs, you should tag @envoyproxy/api-shepherds on any design review.

File: api/envoy/api/v2/cds.proto (hunk 2 above)
Comment (lizan) [reply api/envoy/api/v2/cds.proto]: yes, UpstreamTlsContext in transport_socket.config or typed_config should already work

**Comment by rshriram** [comment]
So @htuch , how about this:

```
cluster:
  tlsContextMap:
    mtls: …, default
    plaintext: …
  endpoints:
    1.1.1.1, tlsContext-to-use: mtls
    2.2.2.2 tlsContext-to-use: plaintext
```

effectively, we ship a bunch of named TLS contexts to the cluster and then in the EDS, we can ship the TLS context to use for a given endpoint. This has the same effect of using label selectors in the TLS context.

**Comment by lizan** [comment]
@rshriram lets iterate on the doc a bit more, I like labels more for its flexibility, a tlscontextmap (or transport socket map) potentially explodes when you have multiple factor to select. (supports mTLS/TLS/cleartext)

**Comment by incfly** [comment]
@lizan not sure if I get it.

The value of the `endpoint.transport_socket_selector` (single hardcoded name) is used to retrieve entries from the map. So this is selection based on single factor, and deterministically, right?

By `label` do you refer to actual `endpoint.metadata` or it does not matter, can be a field in `message Endpoint` as well?

**Comment by incfly** [comment]
PR has been modified to match the proposal, using actual fields in endpoints plus match semantics for transport socket.  @htuch  @lizan  
PTAL, thanks.

File: api/envoy/api/v2/cds.proto [hunk 3]
```diff
@@ -234,6 +235,41 @@ message Cluster {
...
+  //     transport_socket: { "name": "tls", "config": { ... }
+  //   ]
+  // }
+  //
+  // Connection to the endpoints whose "app" label equals to "foo" value uses "enable-mtls"
+  // configuration, a tls socket with detailed configuration specified as *config*.
+  //
+  // If an endpoint can't be matched against any *TransportSockerMatcher*, connection socket
+  // fallbacks to use the *tls_context* or *transport_socket* specified in this cluster.
+  repeated TransportSocketMatcher endpoint_transport_socket_matchers = 41;
```
Comment (lizan) [inline api/envoy/api/v2/cds.proto]: add not-implemented hide annotation, and add a TODO to deprecate transport_socket.

File: api/envoy/api/v2/endpoint/endpoint.proto [hunk 4]
```diff
@@ -74,6 +80,8 @@ message LbEndpoint {
   // to subset the endpoints considered in cluster load balancing.
   core.Metadata metadata = 3;
 
+  repeated Label labels = 6;
```
Comment (lizan) [inline api/envoy/api/v2/endpoint/endpoint.proto]: @htuch

Comment (rshriram) [reply api/envoy/api/v2/endpoint/endpoint.proto]: So @htuch / @lizan I have been giving this some thought. I think its better to have a named TLS context rather than a label selector because named TLS contexts refer to exactly one context to use. Label selectors could select multiple contexts and then we would have to define some first match wins semantics which gets clunky.

Comment (lizan) [reply api/envoy/api/v2/endpoint/endpoint.proto]: I'd like to this be first match wins (similar to FCM), the context map is couples EDS with CDS too tightly, and the matcher semantics here is more flexible to match a subset has two labels etc.

Comment (htuch) [reply api/envoy/api/v2/endpoint/endpoint.proto]: Yeah, let's just have a `string` field here pointing at a specific TLS context. This will be the clearest I reckon.

File: api/envoy/api/v2/cds.proto (hunk 3 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: By deprecation, do you envision there'll be a default transport socket configuration in the list, in order to replace `fallback to outer transport socket/tls_context` field?

I admit that'll be cleaner, just want to confirm.

File: api/envoy/api/v2/endpoint/endpoint.proto (hunk 4 above)
Comment (incfly) [reply api/envoy/api/v2/endpoint/endpoint.proto]: I'm fine with that.

File: api/envoy/api/v2/cds.proto (hunk 3 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: With more thoughts, I think we have to use outer defined `Cluster.transport_socket` can't deprecate it.

This is because for other typed discovery cluster (STRICT_DNS), there's no clear semantics how can they be incorporated into this added fields. It applies only to EDS cluster.

File: api/envoy/api/v2/cds.proto (hunk 2 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: Resolving.

File: api/envoy/api/v2/cds.proto (hunk 3 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: `not-implemented` added.

File: api/envoy/api/v2/endpoint/endpoint.proto (hunk 4 above)
Comment (incfly) [reply api/envoy/api/v2/endpoint/endpoint.proto]: done.

File: api/envoy/api/v2/cds.proto (hunk 3 above)
Comment (lizan) [reply api/envoy/api/v2/cds.proto]: > With more thoughts, I think we have to use outer defined Cluster.transport_socket can't deprecate it.
> This is because for other typed discovery cluster (STRICT_DNS), there's no clear semantics how can they be incorporated into this added fields. It applies only to EDS cluster.

That's another reason I don't like @rshriram's string to transport socket map because it doesn't have default semantics, and much less flexibility for other types.

@rshriram @htuch can we go back to the doc and discuss more, I don't like the string map for the reason above and below.

Comment (incfly) [reply api/envoy/api/v2/cds.proto]: I add a section in the [doc](https://docs.google.com/document/d/1lpshO7E-2KQ5SC519oiRe25TPAXp1HUcyVzZ4QM4IeU/edit?userstoinvite=vladimir%40spotify.com&ts=5d6679a9&actionButton=1#heading=h.3ffxyg8crnt3), and trying to list the pros and cons, PTAL, thanks.

File: api/envoy/api/v2/cds.proto [hunk 5]
```diff
@@ -234,6 +234,26 @@ message Cluster {
...
+  // when the matches conditions are satisfied.
+  message TransportSocketConfig {
+    // The name of the match.
+    string name = 1;
+    // The configuration of the transport socket.
+    core.TransportSocket transport_socket = 3;
+  }
+
+  // Configuration to use different transport sockets for different endpoints.
+  // This should only be used for EDS typed cluster.
```
Comment (rshriram) [inline api/envoy/api/v2/cds.proto]: why only eds? we add the tls object name to the LbEndpoint. Not just eds. There are so many use cases for static clusters

File: api/envoy/api/v2/endpoint/endpoint.proto [hunk 6]
```diff
@@ -74,6 +74,11 @@ message LbEndpoint {
   // to subset the endpoints considered in cluster load balancing.
   core.Metadata metadata = 3;
 
+  // Used as the key to identiy the transport socket configuration defined
+  // in *Cluster.endpoint_transport_socket_map* to use when connecting to this endpoint.
+  // [#not-implemented-hide:]
+  string transport_socket_name = 6;
```
Comment (rshriram) [inline api/envoy/api/v2/endpoint/endpoint.proto]: fyi Istio uses the LbEndpoint for strict dns as well. so this should work there as well right?

File: api/envoy/api/v2/cds.proto (hunk 3 above)
Comment (htuch) [reply api/envoy/api/v2/cds.proto]: Sure, let's take this to the doc.

**Comment by rshriram** [comment]
ping. any updates? Its been two weeks.

File: test/integration/stats_integration_test.cc [hunk 7]
```diff
@@ -252,6 +252,7 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithFakeSymbolTable) {
   // 2019/07/24  7503     43030       44000   add upstream filters to clusters
   // 2019/08/13  7877     42838       44000   skip EdfScheduler creation if all host weights equal
   // 2019/09/02  8118     42830       43000   Share symbol-tables in cluster/host stats.
+  // 2019/09/16  8100     42894       43000   Add transport socket matcher in cluster.
```
Comment (incfly) [inline test/integration/stats_integration_test.cc]: i'm grabbing the number from the failed run of the ci, and trying to adjust these numbers as the comments indicates.

It's intuitively that added field in cluster.prot makes mem footprint larger. I'm not 100% sure this is the right number/changes to made though. 

[1] https://276470-65214191-gh.circle-artifacts.com/0/failed-testlogs/test/integration/stats_integration_test/shard_2_of_2/test.log
[2] https://276470-65214191-gh.circle-artifacts.com/0/failed-testlogs/test/integration/stats_integration_test/shard_1_of_2/test.log

File: api/envoy/api/v2/endpoint/endpoint.proto (hunk 4 above)
Comment (htuch) [reply api/envoy/api/v2/endpoint/endpoint.proto]: Switching back into this review after some time out, it seems that we still have a label match scheme. If the consensus is heading towards this being necessary,  there are two things I'd ask:
1. Document a clear use case of why we have this in `endpoint_transport_socket_matchers` comments.
2. Use the existing `metadata`; I originally pushed back on this as I felt we could just specify a transport socket name here, but now that we are going to a "label" scheme, we already have a source of this information, the endpoint metadata. There should be well-known key name for this, `envoy.transport_socket`.

Comment (incfly) [reply api/envoy/api/v2/endpoint/endpoint.proto]: Done. Added a use case in the end of this field comments. @lizan fyi.

I also renamed the field by removing "endpoint_", since as @rshriram points out, it's possible to add metadata for DNS based endpoint as well.

**Comment by incfly** [comment]
Test are now all passing. PTAL, thanks.

File: api/envoy/api/v2/cds.proto [hunk 8]
```diff
@@ -236,6 +236,65 @@ message Cluster {
...
+  // If an endpoint metadata's value under *envoy.transport_socket* does not match against any
+  // *TransportSockerMatch*, socket configuration fallbacks to use the *tls_context* or
+  // *transport_socket* specified in this cluster.
+  //
+  // This field allows gradual and flexible transport socket configuration changes.
+  // For example, for the same EDS typed cluster, to migrate from plain text traffic to
+  // mutual TLS traffic, management server can configure different Envoy clients with different
+  // *transport_socket_matches*.
+  // [#not-implemented-hide:]
+  repeated TransportSocketMatch transport_socket_matches = 42;
```
Comment (lizan) [inline api/envoy/api/v2/cds.proto]: nit: update `// [#comment:next free field: 42]` to 43, in the line before `message Cluster`

Comment (incfly) [reply api/envoy/api/v2/cds.proto]: Done.

File: api/envoy/api/v2/cds.proto [hunk 9]
```diff
@@ -236,6 +236,65 @@ message Cluster {
...
+  //
+  // Connection to the endpoints whose metadata value under *envoy.transport_socket*
+  // having "app"/"sidecar" key/value pair will use the "enable-mtls" socket configuration.
+  //
+  // If a :ref:`socket match <envoy_api_msg_Cluster.TransportSocketMatch>` with empty match
+  // criteria is provided, that always match any endpoint. For example, the "use-plaintext" socket
+  // configuration in case above.
+  //
+  // If an endpoint metadata's value under *envoy.transport_socket* does not match against any
+  // *TransportSockerMatch*, socket configuration fallbacks to use the *tls_context* or
```
Comment (htuch) [inline api/envoy/api/v2/cds.proto]: Nit: `TransportSocketMatch` .

File: api/envoy/api/v2/cds.proto [hunk 10]
```diff
@@ -236,6 +236,65 @@ message Cluster {
...
+  // :ref:`match <envoy_api_msg_Cluster.TransportSocketMatch>` is used.
+  // For example, with the following match
+  //
+  // .. code-block:: json
+  //
+  // {
+  //   "transport_socket_matches": [
+  //     {
+  //       "name": "enable-mtls",
+  //       "match": { "app": "sidecar" },
```
Comment (htuch) [inline api/envoy/api/v2/cds.proto]: This match example doesn't really show why the label/metadata approach has any benefit. If there is some major win from this (e.g. involving some reduction in combinatorial blow out, etc), can we have an example here that alludes to this?

File: api/envoy/api/v2/cds.proto [hunk 11]
```diff
@@ -236,6 +236,65 @@ message Cluster {
...
+  }
+
+  // Configuration to use different transport sockets for different endpoints.
+  //
+  // The entry of *envoy.transport_socket* in endpoint :ref:`Metadata <envoy_api_msg_core.Metadata>`
+  // is used to match against the transport sockets as they appear in the list. The first
+  // :ref:`match <envoy_api_msg_Cluster.TransportSocketMatch>` is used.
+  // For example, with the following match
+  //
+  // .. code-block:: json
```
Comment (htuch) [inline api/envoy/api/v2/cds.proto]: Nit: please switch to YAML, we tend to prefer this for human readable config fragments as it's less dense with braces.

File: api/envoy/api/v2/cds.proto (hunk 10 above)
Comment (htuch) [reply api/envoy/api/v2/cds.proto]: Maybe the win is around having the same cluster on different Envoys with different transport socket matches, how would this be used in practice? I think it's worth spelling this out here, as IMHO the design is complicated and folks need to know how to use it.

Comment (lizan) [reply api/envoy/api/v2/cds.proto]: @htuch I think what you're asking is expand https://github.com/envoyproxy/envoy/pull/8100/files#diff-1d688ec4c3acf482782efa562b0142c0R293 to actual examples?

If so, I think we need two CDS response examples that explains

- EDS tells "what an endpoint is capable of": i.e. the endpoint has sidecar (or istio) means the endpoint is capable for mTLS and plaintext
- CDS specify how that is being used: CDS for Envoy A can use mTLS for endpoints with sidecar, while CDS for Envoy B keep using plaintext even the endpoints are capable with mTLS.

Comment (incfly) [reply api/envoy/api/v2/cds.proto]: Yes. I stated the reason of having match based sematics at the line  @lizan linked.

The example here originally was trying to illustrate the generic usage of the API. Totally fine to me if we need more concrete example to expand text.  just need to double check. :)

Comment (htuch) [reply api/envoy/api/v2/cds.proto]: Yes, I think just be more verbose. It doesn't even have to be a fully expansion of the CDS response, just paraphrase what @lizan has in those two bullet points.

Comment (incfly) [reply api/envoy/api/v2/cds.proto]: Done. Expand to be more verbose.

File: api/envoy/api/v2/cds.proto (hunk 9 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: done.

File: api/envoy/api/v2/cds.proto (hunk 11 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: done.

**Comment by incfly** [comment]
All tests passed now, PTAL, thanks.

File: api/envoy/api/v2/cds.proto [hunk 12]
```diff
@@ -236,6 +236,82 @@ message Cluster {
...
+  // TransportSocketMatch specifies what transport socket config will be used
+  // when the matches conditions are satisfied.
+  message TransportSocketMatch {
+    // The name of the match.
+    string name = 1;
+
+    // Optional endpoint metadata match criteria.
+    // The connection to the endpoint with metadata matching what is set in this field
+    // will use the transport socket configuration specified here.
+    // The endpoints metadata entry in *envoy.transport_socket* is used to match
```
Comment (incfly) [inline api/envoy/api/v2/cds.proto]: @htuch @lizan 

I'm reading `well_known_names.h`, and see we already have `envoy.transport_sockets.xxx (alts)` for socket name, this makes me think using `envoy.transport_socket` to extract endpoint info can cause confusion.
https://github.com/envoyproxy/envoy/blob/master/source/extensions/transport_sockets/well_known_names.h#L17

how about renaming this to `envoy.transport_socket_match` or `envoy.transport_socket_selector`?

wdty?

File: api/envoy/api/v2/cds.proto [hunk 13]
```diff
@@ -236,6 +236,82 @@ message Cluster {
   //   verification.
   auth.UpstreamTlsContext tls_context = 11;
 
+  // TransportSocketMatch specifies what transport socket config will be used
+  // when the matches conditions are satisfied.
```
Comment (mattklein123) [inline api/envoy/api/v2/cds.proto]: s/matches/match

File: api/envoy/api/v2/cds.proto [hunk 14]
```diff
@@ -236,6 +236,82 @@ message Cluster {
...
+  message TransportSocketMatch {
+    // The name of the match.
```
Comment (mattklein123) [inline api/envoy/api/v2/cds.proto]: Should this be required? What does this do?

File: api/envoy/api/v2/cds.proto (hunk 12 above)
Comment (mattklein123) [inline api/envoy/api/v2/cds.proto]: s/endpoint/endpoint's

File: api/envoy/api/v2/cds.proto [hunk 15]
```diff
@@ -236,6 +236,82 @@ message Cluster {
...
+    // against the values specified in this field.
+    google.protobuf.Struct match = 2;
+
+    // The configuration of the transport socket.
+    core.TransportSocket transport_socket = 3;
+  }
+
+  // Configuration to use different transport sockets for different endpoints.
+  //
+  // The entry of *envoy.transport_socket* in endpoint :ref:`Metadata <envoy_api_msg_core.Metadata>`
```
Comment (mattklein123) [inline api/envoy/api/v2/cds.proto]: Can you actually ref link to the endpoint field itself so this is more clear?

File: api/envoy/api/v2/cds.proto [hunk 16]
```diff
@@ -236,6 +236,82 @@ message Cluster {
...
+  //    match:
+  //      acceptMTLS: true
+  //    transport_socket:
+  //      name: tls
+  //      config: { ... } # tls socket configuration
+  //  - name: "defaultToPlaintext"
+  //    match: {}
+  //    transport_socket:
+  //      name: "rawbuffer"
+
```
Comment (mattklein123) [inline api/envoy/api/v2/cds.proto]: I think this is not going to render correctly due to the empty line. Please do `docs/build.sh` and check the rendered docs.

Or maybe there are now two examples? I'm confused. Can you clean this up into the final form?

File: api/envoy/api/v2/cds.proto [hunk 17]
```diff
@@ -236,6 +236,82 @@ message Cluster {
...
+  //
+  // The metadata of endpoints in EDS can indicate transport socket capabilities. For example,
+  // an endpoint's metadata can have two key value pairs as "acceptMTLS": "true",
+  // "acceptPlaintext": "true". While some other endpoints, only accepting plaintext traffic
+  // has "accpetPlaintext": "true" metadata information.
+  // Then the xDS server can configure the CDS to a client, Envoy A, to send mutual TLS
+  // traffic for endpoints with "acceptMTLS": "true", by adding a corresponding
+  // *TransportSocketMatch* in this field. Other client Envoys receive CDS without
+  // *transport_socket_match* set, and still send plain text traffic to the same cluster.
+  // [#not-implemented-hide:]
```
Comment (mattklein123) [inline api/envoy/api/v2/cds.proto]: This should have arch overview docs once we do the real code change. Please make a note of that.

Comment (incfly) [reply api/envoy/api/v2/cds.proto]: Added a TODO.

File: api/envoy/api/v2/cds.proto (hunk 14 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: Functionality not required given now we're using key/value matching API. But I do plan to use the name as stats/debugging log for troubleshooting. Open to other suggestions as well.

File: api/envoy/api/v2/endpoint/endpoint.proto (hunk 6 above)
Comment (incfly) [reply api/envoy/api/v2/endpoint/endpoint.proto]: right.

File: api/envoy/api/v2/cds.proto (hunk 13 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: done.

File: api/envoy/api/v2/cds.proto (hunk 12 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: done.

File: api/envoy/api/v2/cds.proto (hunk 15 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: done. using envoy_api_field_endpoint.LbEndpoint.metadata for now.

File: api/envoy/api/v2/cds.proto (hunk 16 above)
Comment (incfly) [reply api/envoy/api/v2/cds.proto]: Yeah, it's leftover, cleaned up. I temporaily removed not-implemented-hide to verify the rendering effect.

docs-render

File: api/envoy/api/v2/cds.proto (hunk 14 above)
Comment (mattklein123) [reply api/envoy/api/v2/cds.proto]: That sounds fine. Can we make it required then with a validation?

Comment (incfly) [reply api/envoy/api/v2/cds.proto]: Done. add a min_len = 1, since we don't have required validation for string type per https://github.com/mwitkow/go-proto-validators.

**Comment by incfly** [comment]
@htuch Could you take a look to see anything else we need to update before merging?

File: api/envoy/api/v2/cds.proto [hunk 18]
```diff
@@ -236,6 +236,71 @@ message Cluster {
...
+  // If an endpoint metadata's value under *envoy.transport_socket* does not match any
+  // *TransportSocketMatch*, socket configuration fallbacks to use the *tls_context* or
+  // *transport_socket* specified in this cluster.
+  //
+  // This field allows gradual and flexible transport socket configuration changes.
+  //
+  // The metadata of endpoints in EDS can indicate transport socket capabilities. For example,
+  // an endpoint's metadata can have two key value pairs as "acceptMTLS": "true",
+  // "acceptPlaintext": "true". While some other endpoints, only accepting plaintext traffic
+  // has "accpetPlaintext": "true" metadata information.
```
Comment (htuch) [inline api/envoy/api/v2/cds.proto]: Nit: s/accpetPlaintext/acceptPlaintext/

Comment (incfly) [reply api/envoy/api/v2/cds.proto]: typo fixed.
